src/Grouper.cpp

```diff
@@ -25,191 +25,283 @@
 
 Grouper::Grouper(sortField sf){
   sorter = sf;
-  group_id_i = -1;
-  group_id_set = false;
-  group_id_s = NULL;
-  group_label = NULL;
-  table_index = 1;
-  memset(&stats, 0, sizeof(stats));
+  numGroups = 0;
+  groups = NULL;
 }
 
 /* *************************************** */
 
 Grouper::~Grouper(){
-  if(group_id_s)
-    free(group_id_s);
-  if(group_label)
-    free(group_label);
+  for(int32_t i = 0; i < numGroups; i++) {
+    if(groups[i]) {
+      if(groups[i]->group_id_s)
+        free(groups[i]->group_id_s);
+      if(groups[i]->group_label)
+        free(groups[i]->group_label);
+      free(groups[i]);
+    }
+  }
+  if(groups)
+    free(groups);
 }
 
 /* *************************************** */
 
-bool Grouper::inGroup(Host *h) {
-  if(h == NULL || group_id_set == false)
-    return false;
+/**
+ * Returns group index. Calls newGroup() if no matching group is found.
+ *
+ * Returns -1 for unsupported sorting criteria, causes host to be skipped.
+ */
+int32_t Grouper::inGroup(Host *h) {
+  if(h == NULL)
+    return -1;
 
+  for(int32_t i = 0; i < numGroups; i++) {
```

Comment:
this `for` raises the complexity of the code of a squared factor. Let's say we have N hosts and N groups (worst-case). Function `inGroup` is called for every host and, for every hosts, all the groups are visited with the `for`. So, in the worst-case, the complexity goes from O(N) to O(N^2). It's too much to pay this price.

Note the original idea of class `Grouper`: sort hosts one time according to the group key: then you'll be able to visit them and detect groups automatically simply by checking a change in the group key.

Reply:
BTW, while the added complexity is significant and I accept that it needs to be fixed, it should be kept into account that these changes remove the complexity related to the sorting of the whole set that used to happen in NetworkInterface::getActiveHostsGroup before working on the resulting set in the Grouper. My code only performs one pass on the whole set.

I'm evaluating methods to keep the group list in memory, so as to simplify the search.

Reply:
sorting has a logarithmic complexity, and in the worst-case, the whole set or the groups have an equal number of members so the complexity would still be the same. So you don't have to worry too much if you have to do just 1 sort.